accounts-db/src/io_uring/memory.rs

```diff
@@ -120,49 +130,148 @@ impl DerefMut for PageAlignedMemory {
+        assert!(
+            buffer.len() / FIXED_BUFFER_LEN <= u16::MAX as usize,
+            "buffer too large to register in io_uring"
+        );
+        let buf_start = buffer.as_ptr() as usize;
+
+        buffer.chunks_exact_mut(chunk_size).map(move |buf| {
+            let io_buf_index = (buf.as_ptr() as usize - buf_start) / FIXED_BUFFER_LEN;
+            Self {
+                ptr: buf.as_mut_ptr(),
+                size: buf.len(),
+                io_buf_index: Some(io_buf_index as u16),
+            }
+        })
+    }
+
+    pub fn len(&self) -> usize {
+        self.size
     }
 
     pub fn as_mut_ptr(&self) -> *mut u8 {
         self.ptr
     }
 
-    pub fn len(&self) -> usize {
-        self.size
+    /// The index of the fixed buffer in the ring. See register_buffers().
+    pub fn io_buf_index(&self) -> Option<u16> {
+        self.io_buf_index
     }
 
     /// Return a clone of `self` reduced to specified `size`
-    pub fn sub_buf_to(&self, size: usize) -> Self {
+    pub fn into_shrinked(self, size: usize) -> Self {
         assert!(size <= self.size);
         Self {
             ptr: self.ptr,
             size,
+            io_buf_index: self.io_buf_index,
         }
     }
+
+    /// Registed provided buffer as fixed buffer in `io_uring`.
+    pub fn register_buffer<S, E: RingOp<S>>(
+        ring: &Ring<S, E>,
+        buffer: &mut [u8],
+    ) -> io::Result<()> {
+        adjust_ulimit_memlock(buffer.len())?;
+        let iovecs = buffer
+            .chunks(FIXED_BUFFER_LEN)
+            .map(|buf| libc::iovec {
+                iov_base: buf.as_ptr() as _,
+                iov_len: buf.len(),
+            })
+            .collect::<Vec<_>>();
+        unsafe { ring.register_buffers(&iovecs) }
+    }
+}
+
+impl std::fmt::Debug for IoFixedBuffer {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("IoFixedBuffer")
+            .field("io_buf_index", &self.io_buf_index)
+            .finish()
+    }
 }
 
-impl AsRef<[u8]> for BorrowedBytesMut {
+impl AsRef<[u8]> for IoFixedBuffer {
     fn as_ref(&self) -> &[u8] {
         unsafe { slice::from_raw_parts(self.ptr, self.size) }
     }
 }
+
+impl AsMut<[u8]> for IoFixedBuffer {
+    fn as_mut(&mut self) -> &mut [u8] {
+        unsafe { slice::from_raw_parts_mut(self.ptr, self.size) }
+    }
+}
+
+pub fn adjust_ulimit_memlock(min_required: usize) -> io::Result<()> {
+    let desired_memlock = 2_000_000_000;
+
+    fn get_memlock() -> libc::rlimit {
+        let mut memlock = libc::rlimit {
+            rlim_cur: 0,
+            rlim_max: 0,
+        };
+        if unsafe { libc::getrlimit(libc::RLIMIT_MEMLOCK, &mut memlock) } != 0 {
+            log::warn!("getrlimit(RLIMIT_MEMLOCK) failed");
+        }
+        memlock
+    }
+
+    let mut memlock = get_memlock();
+    let current = memlock.rlim_cur as usize;
+    if current < min_required {
+        memlock.rlim_cur = libc::RLIM_INFINITY;
```

Comment:
hm I find this pretty confusing: we pass a min amount but then we raise to infinity?

Reply:
ah, ok, left-over from attempts to fix CI, it was intended to be set to "desired_memlock", now I updated it and the comment for the constant
```rust
    // This value reflects recommended memory lock limit documented in the validator's
    // setup instructions at docs/src/operations/guides/validator-start.md
    const DESIRED_MEMLOCK: usize = 2_000_000_000;
```